### DISCUSSION TITLE
Singstar Ps3 Microphones

### DISCUSSION THREAD

**Discussion by chaz192** [vicwomg/pikaraoke#275] [General]
Hi, I'm about to install pikaraoke, just wondering if anyone has used Singstar Ps3 Wireless microphones with it, and what sort of setup was required?

Kind regards,

Chaz192

**Comment by slaan1974** [comment]
I have two USB mic's having a delay of at least a second so that kills the fun, looking to see if the Pi has some improvements to get rid of this delay. I guess you have the same with SingStar on PS3 there it works, i have a WII also with SingS using same USB mic's there they work fine. It might be a limitation of the Pi

**Reply by vicwomg** [reply]
You may want to try this: 
https://stackoverflow.com/questions/54452020/raspberry-pi-redirect-usb-microphone-to-hdmi-audio

Particularly:
```
2. Get and compile Tinyalsa https://github.com/tinyalsa/tinyalsa and
   install according to git hub instructions.

...
(need to know the audio card numbers and may also need the card device numbers using -d after the -D's)
  /usr/local/bin/tinycap -- -D <card no> -c 1 -r 44100 -b 16 -p 16 -n 2 | /usr/local/bin/tinyplay - -D <card no> -c 1 -r 44100 -b 16 -p 16 -n 2
```

would be curious to hear your results. Disclaimer: I don't know anything about linux audio

**Comment by vicwomg** [comment]
Ok, as it turns out I have a USB mic and I was able to experiment with this. I was able to achieve pretty low latency USB microphone > headphone out, but NOT hdmi

Install tinyalsa:
```
git clone https://github.com/tinyalsa/tinyalsa.git
cd tinyalsa
make
sudo make install
sudo ldconfig
```

Make note of my audio outputs (card 0 is headphone out, card 1 is hdmi)
```
aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  ...
card 1: vc4hdmi [vc4-hdmi], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

Find my USB microphone card number (card 2, device 0):

```
arecord -l
**** List of CAPTURE Hardware Devices ****
card 2: Device [USB PnP Audio Device], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

Run tinycap and pipe it into tinyplay. Note: "-D 2" argument corresponds to the card number found above. We're piping tincap (which captures the usb mic's output) into tinyplay (immediately play it)

`tinycap -- -D 2 -r 48000 -b 16 -c 1 | tinyplay - -c 1 -r 48000 -b 16 -c 1`

By default tinyplay will output to Card 0, which is the headphone out. This does seem to have decenly low latency (it's noticeable though) . I could not figure out how to get HDMI audio to work, I assume -D 1 would select Card 1 which is perhaps HDMI out?

```
tinycap -- -D 2 -r 48000 -b 16 -c 1 | tinyplay - -D 1 -c 1 -r 48000 -b 16 -c 1
failed to open for pcm 1,0. cannot set hw params: Invalid argument
```

Doesn't seem to like whatever hw params I'm passing to it. I also noticed that tinycap was very sensitive to having the right rate and bits set 48000/16. My USB mic apparently doesn't support 41000/16

**Comment by vicwomg** [comment]
Here is the native way to pipe the USB mic to the default output:
`arecord -D hw:2,0 -f dat -c 1  | aplay -r 48000 -f dat -c 1 -B 1`

But I'm finding the latency unusable. It's under 1s, but it's still not enough

**Comment by slaan1974** [comment]
Thanks for the feedback, give me some time to test and validate it, busy work week ahead

**Comment by slaan1974** [comment]
I tried , but it didn't work out, so i have no other option than to search for a cheap mixer ;)

**Reply by Hartebee5t** [reply]
Look for a wireless mic kit (UHF microphone[s] and receiver) which also supports Bluetooth.  The Pi can send the backing track audio to the mic receiver's Bluetooth channel and then the receiver's microphone volume controls provide the mix levels alongside the Bluetooth level.  The receiver's combined (post-mix) output output can then be fed to an audio amp/speaker.  Better still, find a speaker combo that has Bluetooth and multi-mic inputs.  
Avoid modern digital-based amps though, even if they claim 'Aux In' capability - the Aux In is usually converted via an analog:digital:analog chain that dumps a load more delay that this method otherwise avoids!
[That same conversion latency also rules out most modern TVs as the audio path, hence the suggestion of keeping audio and video paths separate]
